src/modules/mc_pos_control/mc_pos_control_main.cpp

```diff
@@ -1292,784 +1308,797 @@ void MulticopterPositionControl::control_auto(float dt)
-	_arming.armed = false;
+		_pos(0) = _local_pos.x;
+		_pos(1) = _local_pos.y;
 
-	/* get an initial update for all sensor and status data */
-	poll_subscriptions();
 
-	/* We really need to know from the beginning if we're landed or in-air. */
-	orb_copy(ORB_ID(vehicle_land_detected), _vehicle_land_detected_sub, &_vehicle_land_detected);
+		if (_params.alt_mode == 1 && _local_pos.dist_bottom_valid) {
+			_pos(2) = -_local_pos.dist_bottom;
 
-	bool reset_int_z = true;
-	bool reset_int_z_manual = false;
-	bool reset_int_xy = true;
-	bool reset_yaw_sp = true;
-	bool was_armed = false;
+		} else {
+			_pos(2) = _local_pos.z;
+		}
+	}
 
-	hrt_abstime t_prev = 0;
+	if (PX4_ISFINITE(_local_pos.vx) &&
+	    PX4_ISFINITE(_local_pos.vy) &&
+	    PX4_ISFINITE(_local_pos.vz)) {
 
-	math::Vector<3> thrust_int;
-	thrust_int.zero();
+		_vel(0) = _local_pos.vx;
+		_vel(1) = _local_pos.vy;
 
-	// Let's be safe and have the landing gear down by default
-	_att_sp.landing_gear = -1.0f;
+		if (_params.alt_mode == 1 && _local_pos.dist_bottom_valid) {
+			_vel(2) = -_local_pos.dist_bottom_rate;
 
+		} else {
+			_vel(2) = _local_pos.vz;
+		}
+	}
 
-	matrix::Dcmf R;
-	R.identity();
+	_vel_err_d(0) = _vel_x_deriv.update(-_vel(0));
+	_vel_err_d(1) = _vel_y_deriv.update(-_vel(1));
+	_vel_err_d(2) = _vel_z_deriv.update(-_vel(2));
+}
 
-	/* wakeup source */
-	px4_pollfd_struct_t fds[1];
+void
+MulticopterPositionControl::do_position_control(float dt)
+{
 
-	fds[0].fd = _local_pos_sub;
-	fds[0].events = POLLIN;
+	_vel_ff.zero();
 
-	while (!_task_should_exit) {
-		/* wait for up to 20ms for data */
-		int pret = px4_poll(&fds[0], (sizeof(fds) / sizeof(fds[0])), 20);
+	/* by default, run position/altitude controller. the control_* functions
+	 * can disable this and run velocity controllers directly in this cycle */
+	_run_pos_control = true;
+	_run_alt_control = true;
 
-		/* timed out - periodic check for _task_should_exit */
-		if (pret == 0) {
-			// Go through the loop anyway to copy manual input at 50 Hz.
-		}
+	/* select control source */
+	if (_control_mode.flag_control_manual_enabled) {
+		/* manual control */
+		control_manual(dt);
+		_mode_auto = false;
 
-		/* this is undesirable but not much we can do */
-		if (pret < 0) {
-			warn("poll error %d, %d", pret, errno);
-			continue;
+	} else if (_control_mode.flag_control_offboard_enabled) {
+		/* offboard control */
+		control_offboard(dt);
+		_mode_auto = false;
+
+	} else {
+		/* AUTO */
+		control_auto(dt);
+	}
+
+	/* weather-vane mode for vtol: disable yaw control */
+	if (!_control_mode.flag_control_manual_enabled && _pos_sp_triplet.current.disable_mc_yaw_control == true) {
+		_att_sp.disable_mc_yaw_control = true;
+
+	} else {
+		/* reset in case of setpoint updates */
+		_att_sp.disable_mc_yaw_control = false;
+	}
+
+	if (!_control_mode.flag_control_manual_enabled && _pos_sp_triplet.current.valid
+	    && _pos_sp_triplet.current.type == position_setpoint_s::SETPOINT_TYPE_IDLE) {
```

Comment:
I know that this part is just copied from the last one, but I still would like to make a comment for farther refactoring: I think the flag checking such as if idle-state checking needs to happen outside of this function, in contrast to entering a function that does position control and then check if I actually should do it